x-pack/metricbeat/module/gcp/carbon/carbon.go

```diff
@@ -0,0 +1,320 @@
+
+	"github.com/elastic/beats/v7/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+	"github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+)
+
+const (
+	// metricsetName is the name of this metricset
+	metricsetName = "carbon"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet(gcp.ModuleName, metricsetName, New)
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	config config
+	logger *logp.Logger
+}
+
+type config struct {
+	Period              time.Duration `config:"period" validate:"required"`
+	ProjectID           string        `config:"project_id" validate:"required"`
+	CredentialsFilePath string        `config:"credentials_file_path"`
+	CredentialsJSON     string        `config:"credentials_json"`
+	DatasetID           string        `config:"dataset_id" validate:"required"`
+	TableName           string        `config:"table_name"`
+}
+
+// Validate checks for deprecated config options
+func (c config) Validate() error {
+	if c.CredentialsFilePath == "" && c.CredentialsJSON == "" {
+		return errors.New("no credentials_file_path or credentials_json specified")
+	}
+
+	if c.Period.Hours() < 24 {
+		return fmt.Errorf("collection period for carbon footprint metricset %s cannot be less than 24 hours", c.Period)
+	}
+	return nil
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The gcp '%s' metricset is beta.", metricsetName)
+
+	m := &MetricSet{
+		BaseMetricSet: base,
+		logger:        logp.NewLogger(metricsetName),
+	}
+
+	if err := base.Module().UnpackConfig(&m.config); err != nil {
+		return nil, fmt.Errorf("unpack carbon footprint config failed: %w", err)
+	}
+
+	m.Logger().Debugf("metricset config: %v", m.config)
+	return m, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) (err error) {
+	// find current month
+	month := getReportMonth(time.Now())
+
+	var opt []option.ClientOption
```

Comment:
Does this bigquery client setup code from around 96-110 need to be in `Fetch` ? We've shot ourselves in the foot performance-wise before by doing various client setup functions in `Fetch()` instead of `New()`.

Reply:
@endorama do you know why the billing beat does this setup in `Fetch` and not `New`?
I can move this to be initialized once, but I wonder if there isn't something like temporary tokens, or connection that would expire and cause errors when it's only used once every day.

Reply:
One reason I would see is that if we do it in `New`, there's no place where we can call `Close()` on the client.

Reply:
I expect the reason is that given the low frequency to which metricset is expected to run (once every 24h) having a client always ready does not provide benefits.
My suggestion would be to leave it like this unless we are sure it creates a performance penalty.

> I wonder if there isn't something like temporary tokens, or connection that would expire and cause errors when it's only used once every day.

I delved a bit into the inner workings of the BigQuery client, as we use static credentials I don't think the auth issue may happen. The HTTP call to APIs happen in the `Run` method so I don't think terminated connections may happen either. On the other side if we decide to support Application Default Credentials (and I think we should) this may become a problem (and is also difficult to test with such a low frequency).

Reply:
Yah, if it only runs every 24 hours, it's note a huge issue.